### PR TITLE
analyze: make 'blame' work even the default target not reached

### LINKED ISSUE #27187: systemd-analyze blame: coredumps due to Assertion '*a >= b' failed at src/analyze/analyze-time-data.c:15, function subtract_timestamp().
### systemd version the issue has been seen with

253

### Used distribution

Arch

### Linux kernel version used

6.2.2 vanilla

### CPU architectures issue was seen on

x86_64

### Component

systemd-analyze

### Expected behaviour you didn't see

I was expecting the graph of start times.

systemd-analyze by itself (without blame or critical-chain) works:

    # systemd-analyze
    Startup finished in 708ms (userspace)
    initrd.target reached after 638ms in userspace.

### Unexpected behaviour you saw

While working on creating a new target, systemd-analyze coredumps due to a failed assert:

Assertion '*a >= b' failed at src/analyze/analyze-time-data.c:15, function subtract_timestamp(). Aborting

It could be due to missing dependencies (as this is a new target based on minimal dependencies for a dedicated device), but the system works reliably in Wayland, so it's not clear if something (if anything) is missing.

Regardless, in any target mode, systemd-analyze should work to help diagnose issues, and a coredump should only happen for critical issues.

Extra suggestions are welcome to help find the root cause, but I suspect journal entries from previous boots are added to the set of analyzed entries: if that's indeed the case, the error message could show more data, like the precise timestamp or other data for the problematic entry on which the assertion fails.

### Steps to reproduce the problem

systemd-analyze --blame

with the following services reported "running" in systemctl:

init.scope
session-1.scope
bluetooth.service
dbus-broker.service
getty@tty1.service
iwd.service
polkit.service
rtkit-daemon.service
seatd.service
systemd-journald.service
systemd-logind.service
systemd-udevd.service
udisks2.service
upower.service
user@1001.service
dbus.socket
systemd-journald-dev-log.socket
systemd-journald.socket
systemd-udevd-control.socket
systemd-udevd-kernel.socket

And the following services reported active (the running services and .devices have been removed to facilitate legibility)

proc-sys-fs-binfmt_misc.automount
-.mount
dev-hugepages.mount
dev-mqueue.mount
home.mount
run-credentials-systemd\x2dsysctl.service.mount
run-credentials-systemd\x2dtmpfiles\x2dsetup.service.mount
run-credentials-systemd\x2dtmpfiles\x2dsetup\x2ddev.service.mount
run-credentials-systemd\x2dvconsole\x2dsetup.service.mount
run-user-1001-doc.mount
run-user-1001-gvfs.mount
run-user-1001.mount
sys-fs-fuse-connections.mount
sys-kernel-config.mount
sys-kernel-debug-tracing.mount
sys-kernel-debug.mount
sys-kernel-tracing.mount
tmp.mount
kmod-static-nodes.service
lvm2-monitor.service
systemd-backlight@backlight:intel_backlight.service
systemd-backlight@leds:tpacpi::kbd_backlight.service
systemd-modules-load.service
systemd-sysctl.service
systemd-tmpfiles-setup-dev.service
systemd-tmpfiles-setup.service
systemd-udev-trigger.service
systemd-update-utmp.service
systemd-vconsole-setup.service
user-runtime-dir@1001.service
-.slice
system-getty.slice
system-modprobe.slice
system-systemd\x2dbacklight.slice
system-systemd\x2dcoredump.slice
system.slice
user-1001.slice
user.slice
dm-event.socket
lvm2-lvmpolld.socket
systemd-coredump.socket
systemd-rfkill.socket
bluetooth.target
cryptsetup.target
initrd-fs.target
initrd-root-fs.target
integritysetup.target
local-fs.target
network-pre.target
network.target
sound.target
swap.target
sysinit.target
veritysetup.target

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
$ DEBUGINFOD_URLS="https://debuginfod.archlinux.org" gdb --args systemd-analyze blame

Copyright (C) 2023 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-pc-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<https://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from systemd-analyze...

This GDB supports auto-downloading debuginfo from the following URLs:
  <https://debuginfod.archlinux.org>
Enable debuginfod for this session? (y or [n]) y
Debuginfod has been enabled.
To make this setting permanent, add 'set debuginfod enabled on' to .gdbinit.
Downloading separate debug info for /usr/bin/systemd-analyze
Reading symbols from /root/.cache/debuginfod_client/aa5948fc99974013928b5e9b794551879f8a17b2/debuginfo...
(gdb) set print pretty on
(gdb) set pagination off
(gdb) run
Starting program: /usr/bin/systemd-analyze blame
Downloading separate debug info for /lib64/ld-linux-x86-64.so.2
Downloading separate debug info for system-supplied DSO at 0x7ffff7fc8000
Downloading separate debug info for /usr/lib/systemd/libsystemd-core-253.3-1.so
Downloading separate debug info for /usr/lib/systemd/libsystemd-shared-253.3-1.so
Downloading separate debug info for /usr/lib/libseccomp.so.2
Downloading separate debug info for /usr/lib/libc.so.6
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/usr/lib/libthread_db.so.1".
Downloading separate debug info for /usr/lib/libpam.so.0
Downloading separate debug info for /usr/lib/libaudit.so.1
Downloading separate debug info for /usr/lib/libkmod.so.2
Downloading separate debug info for /usr/lib/libmount.so.1
Downloading separate debug info for /usr/lib/libacl.so.1
Downloading separate debug info for /usr/lib/libblkid.so.1
Downloading separate debug info for /usr/lib/libcap.so.2
Downloading separate debug info for /usr/lib/libcrypt.so.2
Downloading separate debug info for /usr/lib/libgcrypt.so.20
Downloading separate debug info for /usr/lib/libip4tc.so.2
Downloading separate debug info for /usr/lib/liblz4.so.1
Downloading separate debug info for /usr/lib/libcrypto.so.3
Downloading separate debug info for /usr/lib/liblzma.so.5
Downloading separate debug info for /usr/lib/libzstd.so.1
Downloading separate debug info for /usr/lib/libm.so.6
Downloading separate debug info for /usr/lib/libdl.so.2
Downloading separate debug info for /usr/lib/libcap-ng.so.0
Downloading separate debug info for /usr/lib/libz.so.1
Downloading separate debug info for /usr/lib/libgpg-error.so.0
Assertion '*a >= b' failed at src/analyze/analyze-time-data.c:15, function subtract_timestamp(). Aborting.

Program received signal SIGABRT, Aborted.
__pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0)
    at pthread_kill.c:44
Downloading source file /usr/src/debug/glibc/glibc/nptl/pthread_kill.c
44            return INTERNAL_SYSCALL_ERROR_P (ret) ? INTERNAL_SYSCALL_ERRNO (ret) : 0;
(gdb) thread apply all bt full

Thread 1 (Thread 0x7ffff6ee3840 (LWP 969314) "systemd-analyze"):
#0  __pthread_kill_implementation (threadid=<optimized out>, signo=signo@entry=6, no_tid=no_tid@entry=0) at pthread_kill.c:44
        tid = <optimized out>
        ret = 0
        pd = <optimized out>
        old_mask = {
          __val = {0}
        }
        ret = <optimized out>
#1  0x00007ffff7886953 in __pthread_kill_internal (signo=6, threadid=<optimized out>) at pthread_kill.c:78
No locals.
#2  0x00007ffff7837ea8 in __GI_raise (sig=sig@entry=6) at ../sysdeps/posix/raise.c:26
        ret = <optimized out>
#3  0x00007ffff782153d in __GI_abort () at abort.c:79
        save_stage = 1
        act = {
          __sigaction_handler = {
            sa_handler = 0x20,
            sa_sigaction = 0x20
          },
          sa_mask = {
            __val = {140737346358883, 1, 140737349971948, 93824992722512, 10091824959817576192, 93824992723048, 93824992354378, 15, 93824992382000, 2, 140737351796480, 18, 140737349876100, 0, 0, 0}
          },
          sa_flags = -136558848,
          sa_restorer = 0x5555555841a0 <times>
        }
#4  0x00007ffff7abcdd7 in log_assert_failed (text=text@entry=0x55555557778c "*a >= b", file=file@entry=0x55555557204a "src/analyze/analyze-time-data.c", line=line@entry=15, func=func@entry=0x555555578c30 <__func__.4.lto_priv.3> "subtract_timestamp") at ../systemd-stable/src/basic/log.c:889
No locals.
#5  0x000055555555efdc in subtract_timestamp (a=<optimized out>, b=<optimized out>) at ../systemd-stable/src/analyze/analyze-time-data.c:15
        __func__ = <optimized out>
#6  subtract_timestamp (a=a@entry=0x5555555c8e10, b=<optimized out>) at ../systemd-stable/src/analyze/analyze-time-data.c:11
        __func__ = "subtract_timestamp"
#7  0x000055555556a28e in acquire_time_data (bus=0x55555558c120, out=out@entry=0x7fffffffdd18) at ../systemd-stable/src/analyze/analyze-time-data.c:270
        t = 0x5555555c8e00
        reply = 0x55555558d3d0
        error = {
          name = 0x0,
          message = 0x0,
          _need_free = 0
        }
        unit_times = 0x5555555c8a10
        boot_times = 0x5555555841a0 <times>
        c = <optimized out>
        u = {
          machine = 0x0,
          id = 0x5555555b3b54 "initrd-root-fs.target",
          description = 0x5555555b3b70 "Initrd Root File System",
          load_state = 0x5555555b3b8c "loaded",
          active_state = 0x5555555b3b98 "active",
          sub_state = 0x5555555b3ba4 "active",
          following = 0x5555555b3bb0 "",
          unit_path = 0x5555555b3bb8 "/org/freedesktop/systemd1/unit/initrd_2droot_2dfs_2etarget",
          job_id = 0,
          job_type = 0x5555555b3bfc "",
          job_path = 0x5555555b3c04 "/"
        }
        r = <optimized out>
        __func__ = "acquire_time_data"
        property_map = {{
            member = 0x555555574970 "InactiveExitTimestampMonotonic",
            signature = 0x555555577d7d "t",
            set = 0x0,
            offset = 16
          }, {
            member = 0x555555577f01 "ActiveEnterTimestampMonotonic",
            signature = 0x555555577d7d "t",
            set = 0x0,
            offset = 24
          }, {
            member = 0x5555555780de "ActiveExitTimestampMonotonic",
            signature = 0x555555577d7d "t",
            set = 0x0,
            offset = 40
          }, {
            member = 0x555555574990 "InactiveEnterTimestampMonotonic",
            signature = 0x555555577d7d "t",
            set = 0x0,
            offset = 32
          }, {
            member = 0x0,
            signature = 0x0,
            set = 0x0,
            offset = 0
          }}
#8  0x000055555556a59c in verb_blame (argc=<optimized out>, argv=<optimized out>, userdata=<optimized out>) at ../systemd-stable/src/analyze/analyze-blame.c:19
        bus = 0x55555558c120
        times = 0x0
        table = 0x0
        cell = <optimized out>
        n = <optimized out>
        r = <optimized out>
        __func__ = "verb_blame"
#9  0x000055555555dea0 in run (argv=0x7fffffffdf18, argc=2) at ../systemd-stable/src/analyze/analyze.c:637
        loop_device = 0x0
        unlink_dir = 0x0
        r = <optimized out>
        loop_device = <optimized out>
        unlink_dir = <optimized out>
        verbs = <optimized out>
        r = <optimized out>
        __func__ = <optimized out>
#10 main (argc=2, argv=0x7fffffffdf18) at ../systemd-stable/src/analyze/analyze.c:640
        r = <optimized out>
        __func__ = "main"
```

### PR DESCRIPTION
**Pull request by yuwata**
Then, this adds a testcase for #27187.

Closes #27187.

### CODE REVIEW

File: test/units/testsuite-01.sh [hunk 1]
```diff
@@ -29,6 +29,12 @@ else
     sysctl -w kernel.core_pattern=dont-overwrite-me
     systemctl daemon-reexec
     diff <(echo dont-overwrite-me) <(sysctl --values kernel.core_pattern)
+
+    # Check that the early setup is actually skipped on reexec.
+    # If the early setup is done more than once, then several timestamps,
+    # e.g. SecurityStartTimestamp, are re-initialized, and causes an ABRT
+    # of systemd-analyze blame. See issue #27187.
+    systemd-analyze blame
```
Comment (mrc0mmand) [inline test/units/testsuite-01.sh]: Could you move this out of the condition, so it runs in nspawn as well?

File: test/units/testsuite-65.sh [hunk 2]
```diff
@@ -11,12 +11,12 @@ export SYSTEMD_LOG_LEVEL=debug
...
 # Sanity checks
 #
-# We can't really test time, blame, critical-chain and plot verbs here, as
+# We can't really test time, critical-chain and plot verbs here, as
 # the testsuite service is a part of the boot transaction, so let's assume
 # they fail
 systemd-analyze || :
 systemd-analyze time || :
-systemd-analyze blame || :
+systemd-analyze blame
```
Comment (mrc0mmand) [inline test/units/testsuite-65.sh]: Could you also add a test for the user daemon? I.e. something like

```
systemd-run --wait --user --pipe -M testuser@.host systemd-analyze blame
```

**Review by mrc0mmand — APPROVED** [review]
Nice, thank you! I had a very similar idea, but using an env var to override the need for a finished bootup to allow it only in tests, but it looks like blame works fine even if the bootup is not yet finished.

**Comment by yuwata** [comment]
@mrc0mmand Thank you for the review. The suggestions are applied.

**Review by YHNdnzj — APPROVED** [review]

**Pull request merged**